server/types/connection.go

```diff
@@ -2,13 +2,21 @@ package types
 
 import (
 	"context"
+	"net"
+	"net/http"
 
 	"github.com/open-telemetry/opamp-go/protobufs"
 )
 
 // Connection represents one OpAMP WebSocket connections.
 // The implementation MUST be a comparable type so that it can be used as a map key.
 type Connection interface {
+	// RemoteAddr returns the remote network address of the connection.
+	RemoteAddr() net.Addr
+
+	// Header returns the request header fields of the http request that opened the connection.
+	Header() http.Header
```

Comment:
Do we need the Header here? The header can be already accessed in the `OnConnecting` callback, so this seems redundant.

Reply:
`OnConnecting` provides the request but at that point there is no agent information available, e.g. `InstanceUid` or `AgentDescription`, which would identify which agent is actually connecting. This information will come later in a message after the connection is established. There is also no `Connection` because the connection isn't established. Providing the Header on the connection will make this information available in the other callbacks.

Reply:
I see. HTTP headers are generally not accessible in gRPC implementations. This presents an issue with the existing `OnConnecting` callback because http.Request is also not available in gRPC implementations. We can take this out if you prefer.

Reply:
Let's add `RemoteAddr()` in this PR and we can think about what we do with `Header()` separately.

Reply:
Done, thanks.

Reply:
For the headers, this is how we handle them for gRPC and HTTP in OpenTelemetry Collector:

https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/configgrpc/configgrpc.go#L401-L409

https://github.com/open-telemetry/opentelemetry-collector/blame/main/config/confighttp/clientinfohandler.go#L52-L59

Reply:
So, looks like it is called Metadata and defined as:
```go
// Metadata is an immutable map, meant to contain request metadata.
type Metadata struct {
	data map[string][]string
}
```